rollbar-android/src/test/java/com/rollbar/android/notifier/sender/ConnectionAwareSenderFailureStrategyTest.java

```diff
@@ -0,0 +1,222 @@
+package com.rollbar.android.notifier.sender;
+
+import com.rollbar.api.payload.Payload;
+import com.rollbar.notifier.sender.SenderFailureStrategy;
+import com.rollbar.notifier.sender.result.Response;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.ArgumentCaptor;
+
+import java.net.ConnectException;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+public abstract class ConnectionAwareSenderFailureStrategyTest {
+  protected ConnectivityDetector connectivityDetector;
+  protected ConnectionAwareSenderFailureStrategy sut;
+  protected Runnable networkRestoredSignal;
+  protected ConnectionAwareSenderFailureStrategy.NanoTimeProvider timeProvider;
+  protected Payload payload;
+
+  @Before
+  public void setUp() {
+    connectivityDetector = mock(ConnectivityDetector.class);
+    ArgumentCaptor<Runnable> signalCaptor = ArgumentCaptor.forClass(Runnable.class);
+
+    timeProvider = mock(ConnectionAwareSenderFailureStrategy.NanoTimeProvider.class);
+    when(timeProvider.nanoTime()).thenReturn(0L);
+
+    sut = new ConnectionAwareSenderFailureStrategy(connectivityDetector, timeProvider);
+    verify(connectivityDetector).setNetworkRestoredSignal(signalCaptor.capture());
+    networkRestoredSignal = signalCaptor.getValue();
+    payload = mock(Payload.class);
+  }
+
+  protected void moveTimeForward(long timeMs) {
+    long current = timeProvider.nanoTime();
+    reset(timeProvider);
+    when(timeProvider.nanoTime()).thenReturn(current + timeMs * 1000000L);
+  }
+
+  public static class BaseTest extends ConnectionAwareSenderFailureStrategyTest {
+    @Test
+    public void shouldNotBeSuspendedAfterInitialization() {
+      assertFalse(sut.isSendingSuspended());
+    }
+
+    @Test
+    public void whenNetworkUpSignalIsCalledSendingShouldResume() {
+      when(connectivityDetector.isNetworkAvailable()).thenReturn(false);
+
+      SenderFailureStrategy.PayloadAction action = sut.getAction(payload,
+          new ConnectException("No connection"));
+
+      assertThat(action, equalTo(SenderFailureStrategy.PayloadAction.CAN_BE_RETRIED));
+      assertTrue(sut.isSendingSuspended());
```

Comment:
A factor to consider here is that `assertEquals` takes the expected value as the first argument, and the actual value as the second argument. You can see how that sometimes confuses people, eg. your suggestion has the args in the wrong order. ;)

The result of this mistake is that if the  assert fails, the message has the opposite meaning to what the developer intended. Eg. "expected 2 got 1" whereas we wanted "expected 1 got 2". 

So I generally prefer `assertThat` for equality, since there's no way to get it wrong (you can't put a matcher object in the first arg.)

`assertThat(someBool, equalTo(true))` is unnecessarily verbose on the other hand, since `assertTrue` has no issues like `assertEquals`. 

But even if `assertEquals` _didn't_ have any issues, I don't see mixing them up as being inconsistent.

Reply:
Good eye! You passed the exercise ;) Yes, hamcrest signatures are reversed compared to jUnit. This is the reason for the so-called inconsistency. Generally speaking, mixing 2 frameworks that are used for the same purpose with different semantic can lead to human errors.

Agree, I prefer to use matchers and `assertThat()` too. My motivation is that you can use more natural language that increases readability and is compatible with both grammar and sentence logic. In fact, that was the reason why the library was created in the first place.

> `assertThat(someBool, equalTo(true))` is unnecessarily verbose on the other hand, since `assertTrue` has no issues like assertEquals.

This is because you are using the wrong matcher here.

You should use `assertThat(someBool, is(true))` which reads as "assert that something is true". You can't go wrong with this semantic.